Adding default User-Agent for netty and okhttp.

nmittler opened the pull request.

nmittler commented:
@ejona86 PTAL

ejona86 commented on netty/src/main/java/io/grpc/transport/netty/Utils.java:
```diff
@@ -72,6 +70,9 @@
   public static final ByteString TE_TRAILERS = new ByteString(HttpUtil.TE_TRAILERS.getBytes(UTF_8));
+  public static final ByteString USER_AGENT = new ByteString("user-agent".getBytes(UTF_8));
+  public static final ByteString DEFAULT_USER_AGENT = new ByteString(
+          "grpc-java-netty".getBytes(UTF_8));
```
Have you thought about adding version? We could add the version to the JAR Manifest as Implementation-Version of io/grpc and then access it here with SomeGrpcClass.class.getPackage().getImplementationVersion()

nmittler replied:
done.

ejona86 commented on core/src/main/java/io/grpc/transport/HttpUtil.java:
```diff
@@ -69,6 +69,13 @@
   public static final String TE_TRAILERS = "trailers";
 
   /**
+   * The User-Agent header named. Defined here since it is not explicitly defined by the HTTP/2
```
I'm not sure I understand this comment. Won't users need to access this? Don't we want it in a more public location?

nmittler replied:
done.

ejona86 commented on netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java:
```diff
@@ -118,45 +128,87 @@ public void teardown() throws Exception {
-      NettyClientTransport transport = new NettyClientTransport(address, NioSocketChannel.class,
-              group, negotiator, DEFAULT_WINDOW_SIZE, DEFAULT_WINDOW_SIZE);
-      transports.add(transport);
+      NettyClientTransport transport = createTransport();
```
This change completely breaks the usefulness of this test. We need the same ProtocolNegotiator used for each transport.

nmittler replied:
oof ... fixed.

ejona86 commented on netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java:
```diff
@@ -118,45 +128,87 @@ public void teardown() throws Exception {
+
+    // Verify that the received headers contained the User-Agent.
+    assertEquals(1, serverListener.streamListeners.size());
+    Metadata.Key<String> userAgentKey = Metadata.Key.of(Utils.USER_AGENT.toString(UTF_8),
```
Why not use HttpUtil.USER_AGENT?

ejona86 replied:
I guess you don't to make sure the "user-agent" strings are the same? Do we want to use HttpUtil.USER_AGENT.name() some places?

nmittler replied:
Just an oversight ... using the common constant for USER_AGENT.

ejona86 commented on netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java:
```diff
@@ -190,7 +242,42 @@ public void onReady() {
     }
   }
 
+  private static final class TestServerStreamListener implements ServerStreamListener {
```
Can we give this a more precise name, like EchoServerStreamListener?

ejona86 commented on netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java:
```diff
@@ -190,7 +242,42 @@ public void onReady() {
+    @Override
+    public void messageRead(InputStream message) {
+      // Just echo back the message.
+      stream.writeMessage(messageStream());
```
To truly echo back, we would need to do `stream.writeMessage(message)`. I see that this was just moved, and it was incorrect before.

nmittler replied:
done.

nmittler commented:
@ejona86 PTAL

nmittler commented:
@ejona86 PTAL.  I've moved the constants under Metadata.Key

nmittler commented:
@ejona86 PTAL ... I've reworked this so that user-agent is a top-level property of `Metadata.Headers` similar to path and authority.

ejona86 commented:
Adding as a first-class property of Headers seems a step back. This does not need to be overridden on a per-call basis. Not allowing the user to override right now is fine with me. If the user could override, it would probably be put in AbstractChannelBuilder.

nmittler commented:
@ejona86 this seems contrary to our previous discussions ... let's talk offline.

nmittler commented:
@ejona86 PTAL. I've removed the changes to `Metadata.Headers` and instead added a `userAgent` option to `AbstractChannelBuilder`. Moved utility method for creating the grpc User-Agent into `HttpUtil`.

ejona86 commented on core/src/main/java/io/grpc/AbstractChannelBuilder.java:
```diff
@@ -87,6 +90,18 @@ public final BuilderT executor(ExecutorService executor) {
   /**
+   * Provides a custom {@code User-Agent} for the application.
+   *
+   * <p>It's an optional parameter. If not provided, a default grpc-specific user agent will be
```
In both cases a grpc-specific user agent is used.

"If provided, the provided user agent will be prepended to grpc's own user agent" or similar?

nmittler replied:
done.

ejona86 commented on core/src/main/java/io/grpc/transport/HttpUtil.java:
```diff
@@ -136,7 +136,7 @@ public static Status httpStatusToGrpcStatus(int httpStatusCode) {
     private final Status status;
 
-    private Http2Error(int code, Status status) {
+    Http2Error(int code, Status status) {
```
Why no longer private?

nmittler replied:
It's redundant with constructors for enums.

ejona86 commented on netty/src/main/java/io/grpc/transport/netty/Utils.java:
```diff
@@ -137,6 +130,14 @@ public static Http2Headers convertClientHeaders(Metadata.Headers headers,
+    String userAgent = GRPC_USER_AGENT;
+    String applicationUserAgent = headers.get(USER_AGENT_KEY);
+    if (applicationUserAgent != null) {
+      userAgent = applicationUserAgent + " " + GRPC_USER_AGENT;
```
In a discussion you weren't a part of, it was decided that the order would be:
language-UA + grpc-core-UA + application-UA

Since we don't have a separate library for language, we wouldn't have language-UA. But it would mean we should invert this concatenation order.

nmittler replied:
done.

nmittler commented:
@ejona86 PTAL

ejona86 commented:
@nmittler LGTM

nmittler commented:
Cherry-picked as 8c1d38a

ywengineer commented:
what does PTAL and LGTM mean?

louiscryan commented:
FYI
PTAL - Please Take Another Look
LGTM - Looks Good To Me
:)


On Wed, Jul 8, 2015 at 6:43 PM, ywengineer <notifications@github.com> wrote:

> what does PTAL and LGTM mean?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-java/pull/508#issuecomment-119779505>.
>

ywengineer commented:
@louiscryan ooooooh, that's interesting. thanks.
